libyul/optimiser/Semantics.h

```diff
@@ -54,11 +54,50 @@ class SideEffectsCollector: public ASTWalker
 		Block const& _ast,
 		std::map<YulString, SideEffects> const* _functionSideEffects = nullptr
 	);
+	SideEffectsCollector(
+		Dialect const& _dialect,
+		ForLoop const& _ast,
+		std::map<YulString, SideEffects> const* _functionSideEffects = nullptr
+	);
 
 	using ASTWalker::operator();
 	void operator()(FunctionCall const& _functionCall) override;
 
 	bool movable() const { return m_sideEffects.movable; }
+
+	bool movableRelativeTo(SideEffects const& _other, bool _codeContainsMSize = true)
+	{
+		if (!m_sideEffects.cannotLoop)
+			return false;
+
+		if (m_sideEffects.movable)
+			return true;
+
+		bool movable = true;
+
+		if (
+			!m_sideEffects.movableApartFromEffects ||
+			m_sideEffects.storage == SideEffects::Write ||
+			m_sideEffects.otherState == SideEffects::Write ||
+			m_sideEffects.memory == SideEffects::Write
+		)
+			movable = false;
```

Comment:
Wouldn't it be easier to just directly return false here (and below)?

Reply:
Fixed all of them to return false. And finally a return true.